File: civ_estimator.py
"""
Dynamic Customer In-Stock Value (CIV / Lambda) Estimator

Estimates how much basket value is lost when an item is out of stock.
No historical transaction data required — uses theoretical estimation from:
  - Velocity (how often item is bought)
  - Demand Stability (planned vs impulse purchase)
  - Category Essentiality (expert prior on basket centrality)
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional
import statistics
import math


@dataclass
class ItemCatalogFeatures:
    """Input: Item characteristics from catalog"""
    asin: str
    category: str
    subcategory: str
    weekly_velocity_units: float  # Mean units sold per week
    demand_cv: Optional[float]    # Coefficient of variation; None if unknown
    price: float


@dataclass
class CIVEstimate:
    """Output: Estimated lambda_ value for Item"""
    asin: str
    lambda_value: float           # The computed CIV ($)
    civ_score: float             # Normalized [0,1] composite before scaling
    velocity_score: float
    stability_score: float
    essentiality_score: float
    confidence: float            # [0,1]; 0 = used fallback prior
    computed_at: datetime
    data_version: str = "v1"


@dataclass
class CIVConfig:
    """Configuration for CIV computation"""
    lambda_min: float = 0.10
    lambda_max: float = 3.00
    weight_velocity: float = 0.35
    weight_stability: float = 0.25
    weight_essentiality: float = 0.40
    fallback_lambda: float = 0.87
    category_essentiality: Dict[str, float] = field(default_factory=lambda: {
        "dairy": 1.00,
        "milk": 1.00,
        "eggs": 0.95,
        "butter": 0.90,
        "household_staples": 0.85,
        "paper_towels": 0.85,
        "laundry": 0.85,
        "pantry_staples": 0.80,
        "oil": 0.80,
        "sugar": 0.78,
        "flour": 0.78,
        "beverages": 0.70,
        "water": 0.70,
        "juice": 0.68,
        "personal_care": 0.65,
        "shampoo": 0.65,
        "toothpaste": 0.65,
        "snacks": 0.45,
        "candy": 0.40,
        "specialty": 0.35,
        "ethnic": 0.35,
        "seasonal": 0.20,
        "discretionary": 0.20,
    })

    # Category-level demand CV priors (when item-level CV unknown)
    category_stability_priors: Dict[str, float] = field(default_factory=lambda: {
        "dairy": 0.08,              # Low CV = stable destination item
        "milk": 0.08,
        "eggs": 0.10,
        "household_staples": 0.12,
        "pantry_staples": 0.15,
        "beverages": 0.18,
        "personal_care": 0.20,
        "snacks": 0.35,             # High CV = impulse purchases
        "seasonal": 0.40,
        "discretionary": 0.45,
    })


class CIVEstimator:
    """Estimates customer in-stock value (lambda_) for inventory items"""

    def __init__(self, config: Optional[CIVConfig] = None):
        self.config = config or CIVConfig()

    def compute_civ(self, features: ItemCatalogFeatures,
                    category_p90_velocity: Optional[Dict[str, float]] = None) -> CIVEstimate:
        """
        Compute CIV estimate for a single item.

        Args:
            features: Item catalog characteristics
            category_p90_velocity: Dict mapping category -> P90 velocity for normalization.
                                  If None, uses a default global P90.

        Returns:
            CIVEstimate with lambda_value and component scores
        """
        if category_p90_velocity is None:
            category_p90_velocity = {}

        # --- Velocity Score: normalized by category P90 ---
        p90_vel = category_p90_velocity.get(features.category.lower(), 50.0)
        if p90_vel > 0:
            velocity_score = min(features.weekly_velocity_units / p90_vel, 1.0)
        else:
            velocity_score = 0.5  # Prior if no category stats
        velocity_score = max(velocity_score, 0.0)

        # --- Stability Score: inverse of demand CV ---
        if features.demand_cv is not None and features.demand_cv >= 0:
            stability_score = 1.0 / (1.0 + features.demand_cv)
        else:
            # Use category-level prior
            category_lower = features.category.lower()
            prior_cv = self.config.category_stability_priors.get(category_lower, 0.25)
            stability_score = 1.0 / (1.0 + prior_cv)

        # --- Essentiality Score: category lookup ---
        category_lower = features.category.lower()
        essentiality_score = self.config.category_essentiality.get(category_lower, 0.45)

        # --- Confidence: how many of the 3 inputs are known ---
        n_known = 0
        if features.weekly_velocity_units > 0:
            n_known += 1
        if features.demand_cv is not None and features.demand_cv >= 0:
            n_known += 1
        if category_lower in self.config.category_essentiality:
            n_known += 1

        confidence = n_known / 3.0

        # --- Composite CIV Score ---
        if confidence < 0.33:
            # Fully unknown item: use prior
            return CIVEstimate(
                asin=features.asin,
                lambda_value=self.config.fallback_lambda,
                civ_score=0.0,
                velocity_score=0.0,
                stability_score=0.0,
                essentiality_score=0.0,
                confidence=0.0,
                computed_at=datetime.now(),
            )

        civ_score = (
            self.config.weight_velocity * velocity_score +
            self.config.weight_stability * stability_score +
            self.config.weight_essentiality * essentiality_score
        )

        # --- Scale to lambda_ range ---
        lambda_value = (
            self.config.lambda_min +
            civ_score * (self.config.lambda_max - self.config.lambda_min)
        )

        return CIVEstimate(
            asin=features.asin,
            lambda_value=lambda_value,
            civ_score=civ_score,
            velocity_score=velocity_score,
            stability_score=stability_score,
            essentiality_score=essentiality_score,
            confidence=confidence,
            computed_at=datetime.now(),
        )

    def batch_compute_civ(self, features_list: List[ItemCatalogFeatures]) -> Dict[str, CIVEstimate]:
        """
        Compute CIV for a batch of items.

        Pre-computes P90 velocity by category for normalization, then processes all items.
        """
        # Compute category-level P90 velocity
        category_velocities: Dict[str, List[float]] = {}
        for feat in features_list:
            cat = feat.category.lower()
            if cat not in category_velocities:
                category_velocities[cat] = []
            if feat.weekly_velocity_units > 0:
                category_velocities[cat].append(feat.weekly_velocity_units)

        category_p90: Dict[str, float] = {}
        for cat, velocities in category_velocities.items():
            if velocities:
                # P90: 90th percentile
                sorted_vel = sorted(velocities)
                idx = min(len(sorted_vel) - 1, int(0.90 * len(sorted_vel)))
                category_p90[cat] = sorted_vel[idx]
            else:
                category_p90[cat] = 50.0  # default

        # Compute CIV for each item
        results: Dict[str, CIVEstimate] = {}
        for feat in features_list:
            est = self.compute_civ(feat, category_p90)
            results[feat.asin] = est

        return results


def demo():
    """Demonstrate CIV estimation with realistic catalog examples"""

    config = CIVConfig()
    estimator = CIVEstimator(config)

    # Create sample catalog
    catalog = [
        # Destination items (high CIV)
        ItemCatalogFeatures("ASIN-001-MILK", "Dairy", "milk", 120.0, 0.08, 3.50),
        ItemCatalogFeatures("ASIN-002-EGGS", "Dairy", "eggs", 95.0, 0.10, 4.20),
        ItemCatalogFeatures("ASIN-003-BUTTER", "Dairy", "butter", 60.0, 0.09, 5.00),

        # Core staples (medium CIV)
        ItemCatalogFeatures("ASIN-004-BREAD", "Pantry", "bread", 85.0, 0.15, 2.50),
        ItemCatalogFeatures("ASIN-005-OIL", "Pantry", "oil", 40.0, 0.12, 8.00),
        ItemCatalogFeatures("ASIN-006-FLOUR", "Pantry", "flour", 35.0, 0.14, 3.00),

        # Household staples
        ItemCatalogFeatures("ASIN-007-TOWELS", "Household", "paper_towels", 70.0, 0.20, 12.00),

        # Low velocity fillers (low CIV)
        ItemCatalogFeatures("ASIN-008-SALT", "Pantry", "salt", 15.0, 0.25, 2.00),
        ItemCatalogFeatures("ASIN-009-SPICE", "Pantry", "spice", 5.0, 0.40, 6.00),
        ItemCatalogFeatures("ASIN-010-CANDY", "Snacks", "candy", 20.0, 0.35, 1.50),
    ]

    # Batch compute
    results = estimator.batch_compute_civ(catalog)

    # Display results
    print("\n" + "=" * 100)
    print("DYNAMIC CIV ESTIMATION - CATALOG ANALYSIS")
    print("=" * 100)
    print(f"{'ASIN':<20} {'Category':<15} {'Velocity':<10} {'λ (CIV)':<10} {'Score':<8} {'Confidence':<12}")
    print("-" * 100)

    for asin in sorted(results.keys()):
        est = results[asin]
        feat = next(f for f in catalog if f.asin == asin)

        print(f"{asin:<20} {feat.category:<15} {feat.weekly_velocity_units:>8.0f} "
              f"${est.lambda_value:>7.2f}   {est.civ_score:>6.2f}  {est.confidence:>6.0%}")

    print("\n" + "=" * 100)
    print("INTERPRETATION")
    print("=" * 100)
    print("λ (CIV) Column:")
    print("  - High λ (~$2.50+): Destination item; OOS loses entire basket")
    print("  - Medium λ (~$0.85): Core item; OOS loses some complementary purchases")
    print("  - Low λ (~$0.20): Filler item; OOS has minimal basket impact")
    print("\nConfidence Column:")
    print("  - 100%: All 3 inputs known (velocity, demand CV, category)")
    print("  - 67%: 2 of 3 known")
    print("  - 33%: 1 of 3 known (uses fallback prior)")
    print("  - 0%: Unknown (uses fallback λ = $0.87)")

    return results


if __name__ == "__main__":
    demo()

File: civ_estimator.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
	"strings"
	"time"
)

// ItemCatalogFeatures represents input item catalog data
type ItemCatalogFeatures struct {
	ASIN                 string
	Category             string
	Subcategory          string
	WeeklyVelocityUnits  float64
	DemandCV             float64 // -1 if unknown
	Price                float64
}

// CIVEstimate represents output: estimated lambda value
type CIVEstimate struct {
	ASIN              string
	LambdaValue       float64
	CIVScore          float64
	VelocityScore     float64
	StabilityScore    float64
	EssentialityScore float64
	Confidence        float64 // [0,1]
	ComputedAt        time.Time
	DataVersion       string
}

// DataSourceConfig specifies where to load event data from
type DataSourceConfig struct {
	Type               string `json:"type"`                 // "file" or "api"
	CheckoutEventsPath string `json:"checkout_events_path"`
	DropoutEventsPath  string `json:"dropout_events_path"`
	APIEndpoint        string `json:"api_endpoint"`
	APIToken           string `json:"api_token"`
}

// FallbackThresholds defines default values when data is unavailable
type FallbackThresholds struct {
	ConfidenceThreshold        float64 `json:"confidence_threshold"`
	DefaultStabilityScore      float64 `json:"default_stability_score"`
	DefaultEssentialityScore   float64 `json:"default_essentiality_score"`
	DefaultP90Velocity         float64 `json:"default_p90_velocity"`
	MinCooccurrenceSupport     int     `json:"min_cooccurrence_support"`
}

// CheckoutEvent represents a completed basket transaction
type CheckoutEvent struct {
	SessionID   string   `json:"session_id"`
	WeekLabel   string   `json:"week_label"` // e.g. "2026-W01"
	BasketASINs []string `json:"basket_asins"`
	Timestamp   string   `json:"timestamp"`
}

// DropoutEvent represents an abandoned basket due to OOS item
type DropoutEvent struct {
	SessionID   string   `json:"session_id"`
	BasketASINs []string `json:"basket_asins"`
	OOSItem     string   `json:"oos_item"`
	Timestamp   string   `json:"timestamp"`
}

// EventLoader interface abstracts loading checkout and dropout events
type EventLoader interface {
	LoadCheckoutEvents() ([]CheckoutEvent, error)
	LoadDropoutEvents() ([]DropoutEvent, error)
}

// FileEventLoader loads events from JSON files
type FileEventLoader struct {
	CheckoutPath string
	DropoutPath  string
}

// APIEventLoader loads events from API (stub for now)
type APIEventLoader struct {
	Endpoint string
	Token    string
}

// ScoreAggregator computes item scores from events
type ScoreAggregator struct{}

// DemoCatalogItem is an intermediate struct for JSON unmarshaling
type DemoCatalogItem struct {
	ASIN                 string  `json:"asin"`
	Category             string  `json:"category"`
	Subcategory          string  `json:"subcategory"`
	WeeklyVelocityUnits  float64 `json:"weekly_velocity_units"`
	DemandCV             float64 `json:"demand_cv"`
	Price                float64 `json:"price"`
}

// CIVConfig holds configuration for CIV computation
type CIVConfig struct {
	LambdaMin          float64            `json:"lambda_min"`
	LambdaMax          float64            `json:"lambda_max"`
	WeightVelocity     float64            `json:"weight_velocity"`
	WeightStability    float64            `json:"weight_stability"`
	WeightEssentiality float64            `json:"weight_essentiality"`
	FallbackLambda     float64            `json:"fallback_lambda"`
	DataSource         DataSourceConfig   `json:"data_source"`
	FallbackThresholds FallbackThresholds `json:"fallback_thresholds"`
	// Runtime-populated maps (not in JSON)
	ItemEssentiality   map[string]float64
	ItemVelocityScore  map[string]float64
	ItemStabilityScore map[string]float64
	ItemAssociations   map[string]map[string]float64
}

// DefaultCIVConfig returns default configuration (no score maps, these are loaded from events)
func DefaultCIVConfig() *CIVConfig {
	return &CIVConfig{
		LambdaMin:          0.10,
		LambdaMax:          3.00,
		WeightVelocity:     0.35,
		WeightStability:    0.25,
		WeightEssentiality: 0.40,
		FallbackLambda:     0.87,
		FallbackThresholds: FallbackThresholds{
			ConfidenceThreshold:      0.33,
			DefaultStabilityScore:    0.80,
			DefaultEssentialityScore: 0.45,
			DefaultP90Velocity:       50.0,
			MinCooccurrenceSupport:   2,
		},
		ItemEssentiality:   make(map[string]float64),
		ItemVelocityScore:  make(map[string]float64),
		ItemStabilityScore: make(map[string]float64),
		ItemAssociations:   make(map[string]map[string]float64),
	}
}

// LoadCheckoutEvents loads checkout events from JSON file
func (f *FileEventLoader) LoadCheckoutEvents() ([]CheckoutEvent, error) {
	data, err := os.ReadFile(f.CheckoutPath)
	if err != nil {
		return nil, err
	}
	var events []CheckoutEvent
	if err := json.Unmarshal(data, &events); err != nil {
		return nil, err
	}
	return events, nil
}

// LoadDropoutEvents loads dropout events from JSON file
func (f *FileEventLoader) LoadDropoutEvents() ([]DropoutEvent, error) {
	data, err := os.ReadFile(f.DropoutPath)
	if err != nil {
		return nil, err
	}
	var events []DropoutEvent
	if err := json.Unmarshal(data, &events); err != nil {
		return nil, err
	}
	return events, nil
}

// LoadCheckoutEvents stub for API loader
func (a *APIEventLoader) LoadCheckoutEvents() ([]CheckoutEvent, error) {
	return nil, fmt.Errorf("API event loader not yet implemented")
}

// LoadDropoutEvents stub for API loader
func (a *APIEventLoader) LoadDropoutEvents() ([]DropoutEvent, error) {
	return nil, fmt.Errorf("API event loader not yet implemented")
}

// Compute aggregates events into item scores using provided thresholds
func (sa *ScoreAggregator) Compute(checkouts []CheckoutEvent, dropouts []DropoutEvent, thresholds FallbackThresholds) (
	essentiality, velocityScore, stabilityScore map[string]float64,
	associations map[string]map[string]float64,
) {
	essentiality = make(map[string]float64)
	velocityScore = make(map[string]float64)
	stabilityScore = make(map[string]float64)
	associations = make(map[string]map[string]float64)

	checkoutCount := make(map[string]int)
	weeklyCount := make(map[string]map[string]int) // asin -> weekLabel -> count
	dropoutCaused := make(map[string]int)
	cooccurrence := make(map[string]map[string]int) // A -> B -> count
	basketCount := make(map[string]int)             // items per basket (for association normalization)

	// Step 1: Build checkout frequency maps
	for _, event := range checkouts {
		for _, asin := range event.BasketASINs {
			checkoutCount[asin]++
			if weeklyCount[asin] == nil {
				weeklyCount[asin] = make(map[string]int)
			}
			weeklyCount[asin][event.WeekLabel]++
		}
		// Count basket-level items for association normalization
		for _, asin := range event.BasketASINs {
			basketCount[asin]++
		}
		// Build co-occurrence matrix
		for i, asinA := range event.BasketASINs {
			if associations[asinA] == nil {
				associations[asinA] = make(map[string]float64)
			}
			for j, asinB := range event.BasketASINs {
				if i != j {
					if cooccurrence[asinA] == nil {
						cooccurrence[asinA] = make(map[string]int)
					}
					cooccurrence[asinA][asinB]++
				}
			}
		}
	}

	// Step 2: Compute velocity scores
	maxCount := 0
	for _, count := range checkoutCount {
		if count > maxCount {
			maxCount = count
		}
	}
	if maxCount > 0 {
		for asin, count := range checkoutCount {
			velocityScore[asin] = float64(count) / float64(maxCount)
		}
	}

	// Step 3: Compute stability scores
	for asin, weekCounts := range weeklyCount {
		if len(weekCounts) == 0 {
			stabilityScore[asin] = 1.0
			continue
		}
		counts := make([]float64, 0)
		sum := 0.0
		for _, count := range weekCounts {
			counts = append(counts, float64(count))
			sum += float64(count)
		}
		mean := sum / float64(len(counts))
		if mean == 0 {
			stabilityScore[asin] = 1.0
			continue
		}
		variance := 0.0
		for _, count := range counts {
			variance += (count - mean) * (count - mean)
		}
		variance /= float64(len(counts))
		stddev := math.Sqrt(variance)
		demandCV := stddev / mean
		stabilityScore[asin] = 1.0 / (1.0 + demandCV)
	}

	// Step 4: Build dropout counts
	for _, event := range dropouts {
		dropoutCaused[event.OOSItem]++
	}

	// Step 5: Compute essentiality scores
	allASINs := make(map[string]bool)
	for asin := range checkoutCount {
		allASINs[asin] = true
	}
	for asin := range dropoutCaused {
		allASINs[asin] = true
	}
	for asin := range allASINs {
		total := dropoutCaused[asin] + checkoutCount[asin]
		if total > 0 {
			essentiality[asin] = float64(dropoutCaused[asin]) / float64(total)
		} else {
			essentiality[asin] = 0.0
		}
	}

	// Step 6: Compute associations (co-occurrence with minimum support from config)
	minSupport := thresholds.MinCooccurrenceSupport
	for asinA, targets := range cooccurrence {
		if associations[asinA] == nil {
			associations[asinA] = make(map[string]float64)
		}
		for asinB, count := range targets {
			if count >= minSupport && basketCount[asinA] > 0 {
				associations[asinA][asinB] = float64(count) / float64(basketCount[asinA])
			}
		}
	}

	return essentiality, velocityScore, stabilityScore, associations
}

// LoadCIVConfig loads configuration from JSON and computes scores from events
func LoadCIVConfig(path string) (*CIVConfig, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	cfg := DefaultCIVConfig()
	if err := json.Unmarshal(data, cfg); err != nil {
		return nil, err
	}

	var loader EventLoader
	switch cfg.DataSource.Type {
	case "file":
		loader = &FileEventLoader{
			CheckoutPath: cfg.DataSource.CheckoutEventsPath,
			DropoutPath:  cfg.DataSource.DropoutEventsPath,
		}
	case "api":
		loader = &APIEventLoader{
			Endpoint: cfg.DataSource.APIEndpoint,
			Token:    cfg.DataSource.APIToken,
		}
	default:
		return nil, fmt.Errorf("unknown data source type: %s", cfg.DataSource.Type)
	}

	checkouts, err := loader.LoadCheckoutEvents()
	if err != nil {
		return nil, fmt.Errorf("failed to load checkout events: %w", err)
	}
	dropouts, err := loader.LoadDropoutEvents()
	if err != nil {
		return nil, fmt.Errorf("failed to load dropout events: %w", err)
	}

	agg := &ScoreAggregator{}
	essentiality, velocityScore, stabilityScore, associations := agg.Compute(checkouts, dropouts, cfg.FallbackThresholds)
	cfg.ItemEssentiality = essentiality
	cfg.ItemVelocityScore = velocityScore
	cfg.ItemStabilityScore = stabilityScore
	cfg.ItemAssociations = associations

	return cfg, nil
}

// CIVEstimator computes customer in-stock value
type CIVEstimator struct {
	Config *CIVConfig
}

// NewCIVEstimator creates a new estimator with default config
func NewCIVEstimator(cfg *CIVConfig) *CIVEstimator {
	if cfg == nil {
		cfg = DefaultCIVConfig()
	}
	return &CIVEstimator{Config: cfg}
}

// ComputeCIV computes CIV for a single item
func (e *CIVEstimator) ComputeCIV(features *ItemCatalogFeatures, categoryP90Velocity map[string]float64) *CIVEstimate {
	if categoryP90Velocity == nil {
		categoryP90Velocity = make(map[string]float64)
	}

	normalizedCategory := strings.ToLower(features.Category)

	// Velocity score: ASIN-keyed lookup first, fall back to P90 method
	var velocityScore float64
	if score, ok := e.Config.ItemVelocityScore[features.ASIN]; ok {
		velocityScore = score
	} else {
		p90 := categoryP90Velocity[normalizedCategory]
		if p90 <= 0 {
			p90 = e.Config.FallbackThresholds.DefaultP90Velocity
		}
		velocityScore = features.WeeklyVelocityUnits / p90
		if velocityScore > 1.0 {
			velocityScore = 1.0
		}
		if velocityScore < 0 {
			velocityScore = 0.0
		}
	}

	// Stability score: ASIN-keyed lookup first, fall back to DemandCV, then default from config
	var stabilityScore float64
	if score, ok := e.Config.ItemStabilityScore[features.ASIN]; ok {
		stabilityScore = score
	} else if features.DemandCV >= 0 {
		stabilityScore = 1.0 / (1.0 + features.DemandCV)
	} else {
		stabilityScore = e.Config.FallbackThresholds.DefaultStabilityScore
	}

	// Essentiality score: ASIN-keyed lookup, fall back to config default
	essentialityScore, ok := e.Config.ItemEssentiality[features.ASIN]
	if !ok {
		essentialityScore = e.Config.FallbackThresholds.DefaultEssentialityScore
	}

	// Confidence: count how many signals are available
	nKnown := 0
	if features.WeeklyVelocityUnits > 0 {
		nKnown++
	} else if _, hasVelocity := e.Config.ItemVelocityScore[features.ASIN]; hasVelocity {
		nKnown++
	}
	if features.DemandCV >= 0 {
		nKnown++
	} else if _, hasStability := e.Config.ItemStabilityScore[features.ASIN]; hasStability {
		nKnown++
	}
	if _, hasEssentiality := e.Config.ItemEssentiality[features.ASIN]; hasEssentiality {
		nKnown++
	}
	confidence := float64(nKnown) / 3.0

	// Handle fully unknown item (confidence below configured threshold)
	if confidence < e.Config.FallbackThresholds.ConfidenceThreshold {
		return &CIVEstimate{
			ASIN:              features.ASIN,
			LambdaValue:       e.Config.FallbackLambda,
			CIVScore:          0.0,
			VelocityScore:     0.0,
			StabilityScore:    0.0,
			EssentialityScore: 0.0,
			Confidence:        0.0,
			ComputedAt:        time.Now(),
			DataVersion:       "v2",
		}
	}

	// Composite CIV score
	civScore := e.Config.WeightVelocity*velocityScore +
		e.Config.WeightStability*stabilityScore +
		e.Config.WeightEssentiality*essentialityScore

	// Scale to lambda range
	lambdaValue := e.Config.LambdaMin + civScore*(e.Config.LambdaMax-e.Config.LambdaMin)

	return &CIVEstimate{
		ASIN:              features.ASIN,
		LambdaValue:       lambdaValue,
		CIVScore:          civScore,
		VelocityScore:     velocityScore,
		StabilityScore:    stabilityScore,
		EssentialityScore: essentialityScore,
		Confidence:        confidence,
		ComputedAt:        time.Now(),
		DataVersion:       "v2",
	}
}

// BatchComputeCIV computes CIV for multiple items
func (e *CIVEstimator) BatchComputeCIV(featuresList []*ItemCatalogFeatures) map[string]*CIVEstimate {
	// Compute category P90 velocities
	categoryVelocities := make(map[string][]float64)
	for _, feat := range featuresList {
		if feat.WeeklyVelocityUnits > 0 {
			normalizedCategory := strings.ToLower(feat.Category)
			categoryVelocities[normalizedCategory] = append(categoryVelocities[normalizedCategory], feat.WeeklyVelocityUnits)
		}
	}

	categoryP90 := make(map[string]float64)
	for cat, velocities := range categoryVelocities {
		if len(velocities) > 0 {
			sort.Float64s(velocities)
			idx := len(velocities) - 1
			if p90Idx := int(0.90 * float64(len(velocities))); p90Idx < len(velocities) && p90Idx > idx {
				idx = p90Idx
			}
			categoryP90[cat] = velocities[idx]
		} else {
			categoryP90[cat] = e.Config.FallbackThresholds.DefaultP90Velocity
		}
	}

	// Compute CIV for each item
	results := make(map[string]*CIVEstimate)
	for _, feat := range featuresList {
		est := e.ComputeCIV(feat, categoryP90)
		results[feat.ASIN] = est
	}

	return results
}

// loadDemoCatalog loads demo catalog from JSON file
func loadDemoCatalog(path string) ([]*ItemCatalogFeatures, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var items []DemoCatalogItem
	if err := json.Unmarshal(data, &items); err != nil {
		return nil, err
	}
	catalog := make([]*ItemCatalogFeatures, len(items))
	for i, item := range items {
		catalog[i] = &ItemCatalogFeatures{
			ASIN:                item.ASIN,
			Category:            item.Category,
			Subcategory:         item.Subcategory,
			WeeklyVelocityUnits: item.WeeklyVelocityUnits,
			DemandCV:            item.DemandCV,
			Price:               item.Price,
		}
	}
	return catalog, nil
}

